plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Scoping.hs

```diff
@@ -0,0 +1,282 @@
+establishScopingBinding regSelf (TypeBind _ (TyVarDecl _ nameDup kind) ty) = do
+    name <- freshenTyName nameDup
+    tyVarDecl <- TyVarDecl (introduceBound name) name <$> establishScoping kind
+    TypeBind NotAName tyVarDecl . referenceVia regSelf name <$> establishScoping ty
+establishScopingBinding regSelf (DatatypeBind dataAnn datatypeDup) = do
+    let Datatype _ dataDeclDup paramsDup matchNameDup constrsDup = datatypeDup
+        TyVarDecl _ dataNameDup dataKind = dataDeclDup
+    dataName <- freshenTyName dataNameDup
+    dataDecl <- TyVarDecl (introduceBound dataName) dataName <$> establishScoping dataKind
+    params <- establishScopingParams paramsDup
+    matchName <- freshenName matchNameDup
+    constrs <- establishScopingConstrsNonRec regSelf dataAnn dataName params constrsDup
+    let datatype = Datatype (introduceBound matchName) dataDecl params matchName constrs
+    pure $ DatatypeBind NotAName datatype
+
+-- | Reference each binding in the last one apart from itself.
+referenceViaBindings
+    :: (forall name. ToScopedName name => name -> NameAnn)
+    -> NonEmpty (Binding TyName Name uni fun NameAnn)
+    -> NonEmpty (Binding TyName Name uni fun NameAnn)
+referenceViaBindings _   (b0 :| [])  = b0 :| []
+referenceViaBindings reg (b0 :| bs0) = go [] b0 bs0 where
+    go prevs b []       = referenceVia reg prevs b :| []
+    go prevs b (c : bs) = b <| go (b : prevs) c bs
+
+-- | Reference each binding in the first one apart from itself and in the last one also apart from
+-- itself. Former bindings are always visible in latter ones and whether latter bindings are visible
+-- in former ones is controlled by the first argument and ultimately depends on the recursivity
+-- of the family of bindings.
+referenceBindingsBothWays
+    :: (forall name. ToScopedName name => name -> NameAnn)
+    -> NonEmpty (Binding TyName Name uni fun NameAnn)
+    -> NonEmpty (Binding TyName Name uni fun NameAnn)
+referenceBindingsBothWays regRec               -- Whether latter bindings are visible in former ones
+    = NonEmpty.reverse                         -- or not depends on the recursivity and so we have
+    . referenceViaBindings regRec              -- the registering function as an argument.
+    . NonEmpty.reverse
+    . referenceViaBindings registerBound       -- Former bindings are always visible in latter ones.
+
+-- | Establish scoping for a family of bindings.
+establishScopingBindings
+    :: (forall name. ToScopedName name => name -> NameAnn)
+    -> NonEmpty (Binding TyName Name uni fun ann)
+    -> Quote (NonEmpty (Binding TyName Name uni fun NameAnn))
+establishScopingBindings regRec =
+    -- Note that mutual recursion and self-recursion are handled separately.
+    fmap (referenceBindingsBothWays regRec) . traverse (establishScopingBinding regRec)
+
+-- | Return a registering function depending on the recursivity.
+registerByRecursivity :: ToScopedName name => Recursivity -> name -> NameAnn
```

Comment:
I'd probably inline this but :shrug:

Reply:
Just tried doing that, but `TypeFamiles` implies [`MonoLocalBinds`](https://downloads.haskell.org/~ghc/8.4.3/docs/html/users_guide/glasgow_exts.html#extension-MonoLocalBinds), which means a local `let`

```haskell
        let regRec = if recy == Rec then registerBound else registerOutOfScope
```

 needs a type signature at which point it's just easier to make a standalone definition.